lib/redux/actions/async.js

```diff
@@ -0,0 +1,560 @@
+    const { devices, uploads, users } = getState();
+    if (uploads.uploadInProgress === true) {
+      return dispatch(syncActions.uploadAborted());
+    }
+
+    const uploadTargetUser = users.uploadTargetUser;
+    dispatch(syncActions.uploadRequest(uploadTargetUser, devices[deviceKey], utc));
+
+    const targetDevice = devices[deviceKey];
+    const deviceType = targetDevice.source.type;
+
+    if (_.includes(['device', 'block'], deviceType)) {
+      dispatch(doDeviceUpload(targetDevice.source.driverId, utc));
+    }
+    else if (deviceType === 'carelink') {
+      dispatch(doCareLinkUpload(deviceKey, opts, utc));
+    }
+  };
+}
+
+export function readFile(userId, deviceKey, file, extension) {
+  return (dispatch, getState) => {
+    if (!file) {
+      return;
+    }
+    dispatch(syncActions.choosingFile(userId, deviceKey));
+    const { version } = getState();
+
+    if (file.name.slice(-extension.length) !== extension) {
+      let err = new Error(errorText.E_FILE_EXT + extension);
+      let errProps = {
+        code: 'E_FILE_EXT',
+        version: version
+      };
+      return dispatch(syncActions.readFileAborted(err, errProps));
+    }
+    else {
+      let reader = new FileReader();
+      reader.onloadstart = () => {
+        dispatch(syncActions.readFileRequest(file.name));
+      };
+
+      reader.onerror = () => {
+        let err = new Error(errorText.E_READ_FILE + file.name);
+        let errProps = {
+          code: 'E_READ_FILE',
+          version: version
+        };
+        return dispatch(syncActions.readFileFailure(err, errProps));
+      };
+
+      reader.onloadend = ((theFile) => {
+        return (e) => {
+          dispatch(syncActions.readFileSuccess(userId, deviceKey, e.srcElement.result));
+          dispatch(doUpload(deviceKey));
+        };
+      })(file);
+
+      reader.readAsArrayBuffer(file);
+    }
+  };
+}
+
+export function doVersionCheck() {
+  return (dispatch, getState) => {
+    dispatch(syncActions.versionCheckRequest());
+    const { api } = services;
+    let { version } = getState();
+    api.upload.getVersions((err, versions) => {
+      if (err) {
+        return dispatch(syncActions.versionCheckFailure(err));
+      }
+      const { uploaderMinimum } = versions;
+      // if either the version from the jellyfish response
+      // or the local uploader version is somehow and invalid semver
+      // we will catch the error and dispatch versionCheckFailure
+      try {
+        const upToDate = semver.gte(version, uploaderMinimum);
+        if (!upToDate) {
+          return dispatch(syncActions.versionCheckFailure(null, version, uploaderMinimum));
+        }
+        else {
+          return dispatch(syncActions.versionCheckSuccess());
+        }
+      }
+      catch(err) {
+        return dispatch(syncActions.versionCheckFailure(err));
+      }
+    });
+  };
+}
+
+/*
+ * COMPLEX ACTION CREATORS
+ */
+
+function getUploadsByUser(targetsByUser) {
```

Comment:
It's used twice (in two thunks), which is why I factored it out. Maybe I'll do a utils file with similar functions I pulled out (`makeUploadCb` etc.).